presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java

```diff
@@ -812,11 +817,25 @@ public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Set<Symb
         @Override
         public PlanNode visitLateralJoin(LateralJoinNode node, RewriteContext<Set<Symbol>> context)
         {
-            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
+            Set<Symbol> expectedFilterSymbols = SymbolsExtractor.extractUnique(node.getFilter());
+
+            Set<Symbol> expectedFilterAndContextSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(expectedFilterSymbols)
+                    .addAll(context.get())
+                    .build();
+
+            PlanNode subquery = context.rewrite(node.getSubquery(), expectedFilterAndContextSymbols);
 
             // remove unused lateral nodes
-            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty() && isScalar(subquery)) {
-                return context.rewrite(node.getInput(), context.get());
+            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty()) {
+                // remove unused lateral subquery of inner join
+                if (node.getType() == INNER && isScalar(subquery) && node.getFilter().equals(TRUE_LITERAL)) {
+                    return context.rewrite(node.getInput(), context.get());
+                }
+                // remove unused lateral subquery of left join
+                if (node.getType() == LEFT && isAtMostScalar(subquery)) {
+                    return context.rewrite(node.getInput(), context.get());
```

Comment:
do you have tests for that? (you can run `TestSubqueries` with coverage and see if this line was executed, then I would like to see focused test for this.

Reply:
There is already a test case which covers removing unused lateral subquery of inner join.
I'm adding another test case for left join.